![Johnson & Johnson MedTech Logo](https://raw.githubusercontent.com/Forage-Simulations/Johnson-Johnson-Robotics-Controls/main/github_assets.png)


# Control System Diagnostic Notebook for Robotic Arm

### Objective
This notebook is designed to help identify and resolve response delays in the control code of a robotic arm. You will:

- Diagnose the root cause of delays in command response times.
- Optimize control code for improved performance.
- Document your findings and propose actionable solutions.

### Instructions
Follow the steps outlined in this notebook to diagnose and resolve issues in the robotic arm's control system:

1. **Load Libraries**: Run the provided setup code to load necessary Python libraries.
2. **Run Diagnostic Functions**: Use the `check_response_time` function to measure command response times.
3. **Analyze Findings**: Record observed delays and propose a hypothesis for their cause.
4. **Test Optimizations**: Apply optimization logic and compare results.
5. **Record Results**: Summarize your findings and recommendations in a structured format.

Focus on the `rotate_joint` command, as it has been flagged for delays.


In [1]:
# Import required libraries
import time  # For measuring response times
import numpy as np  # For numerical calculations


In [2]:
# Define a function to measure the response time of commands
def check_response_time(command):
    """Simulates command execution and measures response time."""
    start_time = time.time()
    if command == "rotate_joint":
        time.sleep(0.1)   #Simulate delay for rotate_joint
    elif command == "move_arm":
        time.sleep(0.1)  # Simulate moderate response time
    elif command == "adjust_grip":
        time.sleep(0.05)  # Simulate fast response time
    response_time = time.time() - start_time
    return response_time


In [3]:
# List of commands to test
commands = ["move_arm", "rotate_joint", "adjust_grip"]

# Measure and print response times for each command
print("Testing initial command response times:")
for cmd in commands:
    response_time = check_response_time(cmd)
    print(f"{cmd} response time: {round(response_time, 3)} seconds")


Testing initial command response times:
move_arm response time: 0.105 seconds
rotate_joint response time: 0.1 seconds
adjust_grip response time: 0.054 seconds


### Step 3: Analyze Initial Findings

Record the response times observed for each command. Focus on identifying commands with higher response times.

| Command       | Observed Response Time | Expected Response Time | Notes on Performance   |
|---------------|------------------------|------------------------|-------------------------|
| move_arm      | 0.105                  | 0.10                   | running as expected     |
| rotate_joint  | 0.155                  | 0.18                   | running better than expected     |
| adjust_grip   | 0.052                  | 0.09                   | running better than expected     |

#### Hypothesis:
The `rotate_joint` command has the longest response time because it has the
longest simulated execution delay, `time.sleep(0.15)`

In [12]:
# Define a function to simulate optimized command execution
def optimized_command(command, improvement_factor=0.2):
    """Simulates optimized command execution."""
    print(f"Optimizing command: {command}")  # Placeholder action
    optimized_response_time = check_response_time(command) * (1 - improvement_factor)
    return optimized_response_time


In [13]:
# Test each command after optimizations
print("\nTesting optimized command response times:")
for cmd in commands:
    optimized_time = optimized_command(cmd)
    print(f"{cmd} optimized response time: {round(optimized_time, 3)} seconds")



Testing optimized command response times:
Optimizing command: move_arm
move_arm optimized response time: 0.083 seconds
Optimizing command: rotate_joint
rotate_joint optimized response time: 0.121 seconds
Optimizing command: adjust_grip
adjust_grip optimized response time: 0.043 seconds


### Step 5: Record Results

Use this section to document your findings from the optimization process.

#### Observations:
- Initial Response Times:
  - move_arm: 0.105
  - rotate_joint: 0.155
  - adjust_grip: 0.052

- Optimized Response Times:
  - move_arm: 0.084
  - rotate_joint: 0.123
  - adjust_grip: 0.043

#### Key Insights:
- Which commands showed improvement after optimization?
    All three commands showed an estimated 20% improvement. `rotate_joint' showed the largest absolute reduction because it initially had the longest response time.

- Were any specific optimizations particularly effective?
    The improvement factor estimated a 20% reduction for every command. However, it reduced the reported measurement mathematically rather than changing the command's actual execution delay.


### Step 6: Summary and Recommendations

Summarize your findings, including identified issues, optimizations applied, and next steps.


**Identified Issue:** The `rotate_joint` command had the longest measured
response time because it used the longest simulated execution delay.

**Optimization Applied:** A 20% improvement factor was applied to estimate
the potential optimized response time.


**Next Steps:**
- Further refine code for additional improvements.
- Conduct stress testing to validate performance under high-load scenarios.
